Compiler for Code Switching

inctechs opened the pull request:
## Description

Compiler for determining the minimum number of code switching operations.

## Checklist:

<!---
This checklist serves as a reminder of a couple of things that ensure your pull request will be merged swiftly.
-->

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [ ] I have added entries to the changelog for any noteworthy additions, changes, fixes, or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [ ] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

inctechs committed:
update git_ignore

inctechs committed:
Merge remote-tracking branch 'origin/main' into code-switching-compiler

inctechs committed:
add networkx to mypy ignore

inctechs committed:
implement first version of min-cut solution

inctechs committed:
add one way transveral CNOTs to be allowed

For this we fixed the arbitrary convention that the 2D Code corresponds
to the source and the 3D code to the sink.

inctechs committed:
add edge_type keyword to edges

this helps to distinguish between unary edges used for bias and temporal
edges used for actual switching locations

inctechs committed:
change default order of magnitude for temporal edges

inctechs committed:
add code bias functionality

inctechs committed:
clear/ reorder code

inctechs committed:
read in QC as dag to support depth and prepare idle switching

inctechs committed:
change qubit activity collection to list instead of set

inctechs committed:
make Global default temporal edge weight

inctechs committed:
add idle bonus calculation and idle switching funcitonality

inctechs committed:
update docstrings and comments

inctechs committed:
fix counting of switches

prior every edge with nodes in different codes were counted but this
included one-way

inctechs committed:
Revert "fix counting of switches"

This reverts commit 73ffd1c36c4349104c8a4fb628370fed062627ee.

inctechs committed:
fix counting of switching operations

now infinite edges are of type "fixed" and won't be count

inctechs committed:
add placeholder switch operations to qiskit circuit

now we can add to a given qiskit circuit the switching operations as
placeholder operations with a given depth to simulate how much switching
operations increase circuit depth

inctechs committed:
add extraction of switching position

inctechs committed:
add feature that idle bonus only considered for certain length

bonus will only be applied when the idle lenght is longer than the
assumed switching length in units of single qubit lengths

inctechs committed:
change return of min-cut function

inctechs committed:
add scripts for generating ciruits

inctechs committed:
add function to generate universal random qiskit circuit

inctechs committed:
add edge ration

now difference betweem biased (old unary) edges and temporal edges is
given by a ration

inctechs committed:
change default folder name of circuit output

inctechs committed:
update default number of seeds

inctechs committed:
add lookahead function

inctechs committed:
add scripts for performance analysis

inctechs committed:
change default number of generated cicuits

inctechs committed:
remove raw data return form min_cut

inctechs committed:
make lookahead deterministic

inctechs committed:
add lookahead to init file

inctechs committed:
change simulations to be gate distribution sensitive

inctechs committed:
add feature to avoid generating circuits twice

inctechs committed:
add script to generate array with steps of 64

inctechs committed:
add all qubits numbers to performance bash

inctechs committed:
add intermediate results

inctechs committed:
add new results

inctechs committed:
update simulations

inctechs committed:
move utils functions to utils file

inctechs committed:
set up code switching compilation test

inctechs committed:
Merge remote-tracking branch 'origin/main' into code-switching-compiler

inctechs committed:
repair test for idle bonus

inctechs committed:
redefine api by making functions private

inctechs committed:
add full function docstring extract switch locations

inctechs committed:
add one way transversal test

inctechs committed:
add CompilerConfig for different settings and keep it tidy

inctechs committed:
add CompilerConfig to init

inctechs committed:
connect biased code to CompilerConfig

inctechs committed:
add code bias test

inctechs committed:
add test for compilation utils

inctechs committed:
remove dag layer visualization used for debugging previously

inctechs committed:
Merge remote-tracking branch 'origin/main' into code-switching-compiler

inctechs committed:
work in CodeRabbitAI code review

inctechs committed:
fix more CodeRabbitAIs comments

pehamTom commented on pyproject.toml:
```diff
@@ -153,7 +153,7 @@ exclude = [
 module = ["qiskit.*", "qecsim.*", "qiskit_aer.*", "matplotlib.*", "scipy.*", "ldpc.*", "pytest_console_scripts.*",
-    "z3.*", "bposd.*", "numba.*", "pymatching.*", "stim.*", "multiprocess.*", "sinter.*", "qsample.*", "pandas.*", "tqdm.*"]
+    "z3.*", "bposd.*", "numba.*", "pymatching.*", "stim.*", "multiprocess.*", "sinter.*", "qsample.*", "pandas.*", "tqdm.*", "networkx.*"]
 ignore_missing_imports = true
```
I think the bot is right in this case.

pehamTom commented on scripts/cs_compiler/run_generate_circuits.sh:
```diff
@@ -0,0 +1,32 @@
+    local n=$1
+    local distr_type=$2
+    python "${SCRIPT_DIR}/generate_random_circuits.py" --n "$n" --num_circuits "$num_circuits"
+}
```
Also seems true.

pehamTom commented on scripts/cs_compiler/simulate_circuit_performance.py:
```diff
@@ -0,0 +1,109 @@
+    with csv_path.open("r", newline="", encoding="utf-8") as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            # safe integer comparison
```
```suggestion
```

pehamTom commented on src/mqt/qecc/circuit_compilation/code_switching_compiler.py:
```diff
@@ -0,0 +1,369 @@
+            idle_length = 0
+
+        max_bonus = 5
+        return min(max_bonus, idle_length) * self.base_unary_capacity
```
This is not what we describe in the paper and does not necessarily preserve minimality. Or am I missing something?

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+    return circuit
+
+
+def count_code_switches(circuit: QuantumCircuit) -> tuple[int, list[str | None]]:
```
This is the naive method, right? Maybe just call it `naive_switching` or something?

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+        gate = instr.operation.name
+        qubits = [circuit.find_bit(q).index for q in instr.qubits]
+
+        # Skip ID gates, they don't constrain anything
```
```suggestion
```

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+            if current_code[q] is None:
+                # Assign to one of the compatible codes (non-deterministically -> set has no order)
+                # current_code[q] = next(iter(compat))
+                # deterministic choice for testing
```
```suggestion
```

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+pattern = re.compile(r".*_q(\d+)_d(\d+)")
+
+
+def parse_node_id(node_id: str) -> tuple[int, int]:
```
I don't understand why this is not just in the other file since this only makes sense in combination with the Graph.

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+    # Iterate layers in order; copy each op, then after the whole layer insert placeholders for that layer
+    for depth_idx, layer in enumerate(layers):
+        layer_graph = layer["graph"]
+        # append ops of the layer to new_qc (deterministic order: iterate nodes)
```
```suggestion
```

pehamTom commented on src/mqt/qecc/circuit_compilation/compilation_utils.py:
```diff
@@ -0,0 +1,293 @@
+            assert isinstance(node, DAGOpNode)
+            # Map node.qargs (Qubit objects) to the corresponding qubit objects of the original circuit
+            q_indices = [circuit.find_bit(q).index for q in node.qargs] if node.qargs else []
+            qbit_objs = [circuit.qubits[i] for i in q_indices]
```
Honestly don't know but doesn't `find_bit` already give you the qubit object?

pehamTom commented on tests/circuit_compilation/test_code_switching_compilation.py:
```diff
@@ -0,0 +1,172 @@
+
+    # Should flow entirely through Source
+    assert len(switch_pos) == 0
+    assert num_switches == 0.0
```
Shouldn't `num_switches` always be an int?

pehamTom commented on tests/circuit_compilation/test_code_switching_compilation.py:
```diff
@@ -0,0 +1,172 @@
+
+    # Now, no switches should be needed
+    assert num_switches == 0
+    assert switch_pos == []  # Switch on qubit 0 at depth 0
```
Nit: above you tested `len(switch_pos)==0`

pehamTom commented on tests/circuit_compilation/test_code_switching_compilation.py:
```diff
@@ -0,0 +1,172 @@
+        # Ensure Source is in S and Sink is in T
+        assert csg.source in S
+        assert csg.sink in T
+
```
Bot is right.

pehamTom commented on tests/circuit_compilation/test_compilation_utils.py:
```diff
@@ -0,0 +1,162 @@
+def test_count_switches_invalid_gate():
+    """Test that an unsupported gate raises ValueError."""
+    qc = QuantumCircuit(1)
+    qc.x(0)  # 'x' is not in the defined sets for Code A or B in the utils
```
In the context of code switching somewhat of a strange test, since both 2D and 3D color code support an X gate. I know why you do it like this but just a bit odd when I stumbled over it.

pehamTom reviewed (changes requested):
Already looks really good. I think some things can just be cleaned up. My biggest comment would be to just move everything in src into one file. The utils are very specific to the Graph class or can be moved to scripts.

inctechs committed:
rename variables in scripts

inctechs committed:
rename class to MinimalCodeSwitchingCompiler

inctechs committed:
generalise Graph construction

before we hard coded 2D and 3D color codes with only H,T,CX gates. Now
when initialising the Graph we require that the two transversal gate
sets are provided. And based on that we create the graph. This allows
for different structures.

inctechs committed:
remove unused methods to connect to terminals and connecting CNOTs

inctechs committed:
fix typo with parameter names

inctechs committed:
add idle bonus calcualtion from the paper

inctechs committed:
move random circuit generation from utils to scripts

inctechs committed:
rename naive code switching method and adjust collaterals

inctechs committed:
remove unnecessary comments

inctechs committed:
move node parser back to graph class

inctechs committed:
remove redundancy

inctechs committed:
rename tests for naive switching

inctechs committed:
remove stress test, adapt parsing node test, fix num_switches typo

inctechs committed:
improve placeholder test

inctechs committed:
remove personal playground folder

inctechs committed:
add networkx to pyproject.toml

inctechs committed:
automatic reordering by uv

inctechs committed:
fix logical error where same back to back gates are allowed

inctechs committed:
rabbit edits

inctechs committed:
rename package to code_switching

inctechs committed:
add safe guard for non negative bonus

inctechs committed:
save uv.lock

inctechs committed:
Merge remote-tracking branch 'origin/main' into code-switching-compiler

inctechs committed:
add CodeSwitching docs

inctechs committed:
rename unary edges to bias edges

inctechs committed:
generalise one-way transversal gate definition.

Definition should state which gates are one-way transversal and also in
which direction this one-way-character holds.

inctechs committed:
sketch doc structure

inctechs committed:
fix bug in docs

inctechs committed:
fix bug in code cells

inctechs committed:
fix typos in the docs

pehamTom reviewed (approved):
LGTM, gonna approve already but would be great if you could look at my comments for the docs.

Overall, nice work!

inctechs committed:
Merge remote-tracking branch 'origin/main' into code-switching-compiler

inctechs committed:
Update docs/CodeSwitching.md

Co-authored-by: Tom Peham <tom.peham@tum.de>
Signed-off-by: Erik Weilandt <48921674+inctechs@users.noreply.github.com>

inctechs committed:
Apply suggestions from code review for docs

Co-authored-by: Tom Peham <tom.peham@tum.de>
Signed-off-by: Erik Weilandt <48921674+inctechs@users.noreply.github.com>

inctechs committed:
Apply suggestions from code review: fix naming of class in docstrings

Co-authored-by: Tom Peham <tom.peham@tum.de>
Signed-off-by: Erik Weilandt <48921674+inctechs@users.noreply.github.com>

inctechs committed:
edit Changelog.md

inctechs committed:
Merge branch 'code-switching-compiler' of https://github.com/munich-quantum-toolkit/qecc into code-switching-compiler

inctechs commented on src/mqt/qecc/code_switching/code_switching_compiler.py:
```diff
@@ -0,0 +1,473 @@
+
+        prev_depth, curr_depth = depths[-2], depths[-1]
+        bonus = self.compute_idle_bonus(prev_depth, curr_depth, total_edges)
+        return max(0.0, base_capacity - bonus)
```
jap aber code rabbit hat mich 10x genervt, dass ich machen soll als backup

inctechs committed:
fix latest coderabbit feedback

inctechs committed:
add missing references to changelog

inctechs committed:
Merge remote-tracking branch 'origin' into code-switching-compiler

pre-commit-ci[bot] committed:
🎨 pre-commit fixes

inctechs commented:
@coderabbitai pause

inctechs committed:
fix last coderabbit changes

inctechs committed:
Merge branch 'code-switching-compiler' of https://github.com/munich-quantum-toolkit/qecc into code-switching-compiler

inctechs committed:
remove note from README and add key feature description
